File: csharp/rational-numbers/RationalNumbers.cs
using System;
using System.Diagnostics;

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    {
        return Math.Pow((double)realNumber, (double)r.num / (double)r.den);
    }
}

public struct RationalNumber
{
    public int num;
    public int den;
    public RationalNumber(int numerator, int denominator)
    {
        num = numerator;
        den = denominator;
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        if (r1.den == r2.den) return new RationalNumber(r1.num + r2.num, r1.den).Reduce();
        return new RationalNumber(r1.num * r2.den + r2.num * r1.den, r1.den * r2.den).Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        if (r1.den == r2.den) return new RationalNumber(r1.num - r2.num, r1.den).Reduce();
        return new RationalNumber(r1.num * r2.den - r2.num * r1.den, r1.den * r2.den).Reduce();
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        return new RationalNumber(r1.num * r2.num, r1.den * r2.den).Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        return (r1 * Inverse(r2)).Reduce();
    }

    public static RationalNumber Inverse(RationalNumber r)
    {
        return new RationalNumber(r.den, r.num);
    }

    public RationalNumber Abs()
    {
        return new RationalNumber(Math.Abs(num), Math.Abs(den));
    }

    public RationalNumber Reduce()
    {
        if (num == 0 || den == 1) return new RationalNumber(num, 1);
        int a = num;
        int b = den;
        while (a != 0)
        {
            int c = a;
            a = b % a;
            b = c;
        }
        if (b == 1) return this;
        if ( (b < 0 && den < 0) || (b > 0 && den > 0) ) return new RationalNumber(num / b, den / b);
        return new RationalNumber( - num / b, - den / b);
    }

    public RationalNumber Exprational(int power)
    {
        if (power == 0) return new RationalNumber(1, 1);
        if (power > 0) return new RationalNumber((int)Math.Pow((double)num, power), (int)Math.Pow((double)den, power)).Reduce();
        else return Inverse(Exprational(-power));
    }

    public double Expreal(int baseNumber)
    {
        return Math.Pow(baseNumber, ((double)num) / ((double)den));
    }
}

File: fsharp/rational-numbers/RationalNumbers.fs
module RationalNumbers

open System

[<Struct>]
type Fraction = { Num: int; Den: int }

let rec gcd a b = if b = 0 then a else gcd b (a % b)

let create numerator denominator = 
    let g = gcd numerator denominator
    if (g > 0 && denominator > 0) || (g < 0 && denominator < 0) then
        { Num = numerator / g; Den = denominator / g}
    else
        { Num = - numerator / g; Den = - denominator / g }

let add (r1: Fraction) (r2: Fraction) = create (r1.Num * r2.Den + r1.Den * r2.Num) (r1.Den * r2.Den)

let sub (r1: Fraction) (r2: Fraction) = create (r1.Num * r2.Den - r1.Den * r2.Num) (r1.Den * r2.Den)

let mul (r1: Fraction) (r2: Fraction) = create (r1.Num * r2.Num) (r1.Den * r2.Den)

let inv (r: Fraction) = create (r.Den) (r.Num)

let div (r1: Fraction) (r2: Fraction) = mul r1 (inv r2)

let abs (r: Fraction) = create (abs (r.Num)) (abs (r.Den))

let exprational n (r: Fraction) = create (int <| Math.Pow(double r.Num, double n)) (int <| Math.Pow(double r.Den, double n))

let expreal (r: Fraction) n = Math.Pow(double n, (double (r.Num) / double (r.Den)))

let reduce (r: Fraction) = create r.Num r.Den